tensorly/regression/cp_plsr.py

```diff
@@ -0,0 +1,230 @@
+from ..cp_tensor import CPTensor, cp_to_tensor, cp_to_vec
+from ..decomposition import tucker
+from .. import backend as T
+from .. import unfold
+
+# Author: Cyrillus Tan, Jackson Chin, Aaron Meyer
+
+# License: BSD 3 clause
+
+
+class CP_PLSR:
+    """CP tensor regression
+
+        Learns a low rank CP tensor weight
+
+    Parameters
+    ----------
+    n_components : int
+        rank of the CP decomposition of the regression weights
+    tol : float
+        convergence value
+    n_iter_max : int, optional, default is 100
+        maximum number of iteration
+    random_state : None, int or RandomState, optional, default is None
+    verbose : bool, default is False
+        whether to be verbose during fitting
+    """
+
+    def __init__(
+        self, n_components, tol=1.0e-7, n_iter_max=100, random_state=None, verbose=False
+    ):
+        self.n_components = n_components
+        self.tol = tol
+        self.n_iter_max = n_iter_max
+        self.random_state = random_state
+        self.verbose = verbose
+
+    def get_params(self, **kwargs):
+        """Returns a dictionary of parameters"""
+        params = ["n_components", "tol", "n_iter_max", "random_state", "verbose"]
+        return {param_name: getattr(self, param_name) for param_name in params}
+
+    def set_params(self, **parameters):
+        """Sets the value of the provided parameters"""
+        for parameter, value in parameters.items():
+            setattr(self, parameter, value)
+        return self
+
+    def fit(self, X, Y):
+        """Fits the model to the data (X, Y)
+
+        Parameters
+        ----------
+        X : ndarray
+            tensor data of shape (n_samples, N1, ..., NS)
+        Y : 2D-array of shape (n_samples, n_predictions)
+            labels associated with each sample
+
+        Returns
+        -------
+        self
+        """
+        ## PREPROCESSING
+        # Check that both tensors are coupled along the first mode
+        if T.shape(X)[0] != T.shape(Y)[0]:
+            raise ValueError(
+                "The first modes of X and Y must be coupled and have the same length."
+            )
+        X, Y = T.copy(X), T.copy(Y)
+
+        # Check the shape of X and Y; convert vector Y to a matrix
+        if T.ndim(X) < 2:
+            raise ValueError("X must be at least a 2-mode tensor.")
+        if (T.ndim(Y) != 1) and (T.ndim(Y) != 2):
+            raise ValueError("Only a matrix (2-mode tensor) Y is allowed.")
+        if T.ndim(Y) == 1:
+            Y = T.reshape(Y, (-1, 1))
+
+        # Mean center the data, record info the object
+        self.X_shape = T.shape(X)
+        self.Y_shape = T.shape(Y)
+        self.X_mean = T.mean(X, axis=0)
+        self.Y_mean = T.mean(Y, axis=0)
+        X -= self.X_mean
+        Y -= self.Y_mean
+
+        self.X_factors = [T.zeros((l, self.n_components)) for l in T.shape(X)]
+        self.Y_factors = [
+            T.tile(Y[:, [0]], self.n_components),
```

Comment:
It would be better to implement this as a list comprehension and then v/h stack. I know that's weird, but tile seems to work differently in PyTorch which is causing issues.